src/FirmwarePlugin/APM/APMFirmwarePlugin.cc

```diff
@@ -360,14 +357,10 @@ void APMFirmwarePlugin::adjustOutgoingMavlinkMessageThreadSafe(Vehicle* vehicle,
 
 QString APMFirmwarePlugin::_getMessageText(mavlink_message_t* message) const
 {
-    QByteArray b;
-
-    b.resize(MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1);
-    mavlink_msg_statustext_get_text(message, b.data());
-
-    // Ensure NUL-termination
-    b[b.length()-1] = '\0';
-    return QString(b);
+    mavlink_statustext_t data;
+    mavlink_msg_statustext_decode(message, &data);
+    QString text = QString::fromStdString(data.text);
```

Comment:
I disagree here. The previous implementation was there for a reason. `data.text` is not necessarily 0 terminated by spec, so this can read garbage memory.

What about using [QString::fromLocal8Bit()](https://doc.qt.io/qt-6/qstring.html#fromLocal8Bit) which allows to pass the max length as well?

Reply:
I'm sure I had a reason, but I don't remember why I even touched this part in this PR. I'll have to go back and check.

Reply:
So maybe this and the other fix should be separate PRs, but the reason I did this was because when changing the QRegExp in _handleIncomingStatusText, I noticed it was impossible for _coaxialMotors to be set as true because of the trailing 0's, as shown in this image. Maybe it's better to just check if the QStringLiterals are substrings of frameType.
![image](https://github.com/mavlink/qgroundcontrol/assets/68555040/952738f6-a063-496c-865c-78ad59baa060)

Reply:
I'm confused. A statustext is a string closed with `\0` or not if the full length is used. If you construct QString from it, it will read past the array into garbage and we need to protect against that. I'm not sure how that matters with a regexp? Also, this looks like statustext is potentially somewhat abused for this use case :thinking:?

Reply:
So it's not really related to QRegExp, which is why I probably should have made a separate issue/PR, but I tried to fix it when I was changing this code anyways. The problem the way things are is that you cannot do an equality comparison of `frameType` (which is always 50 bytes) directly to "Y6", "OCTA_QUAD", etc. That equality will never be true. So something needs to be changed with the `frameType` comparison, either by dropping the trailing zeros or using substring or something like that.

Reply:
Right, what about if you convert both to QString and then compare? That would drop trailing 0s, right?

Reply:
They already are both QStrings. The thing is in `_getMessageText`, the `\0` termination is always added at the very end. So you will always have a bunch of trailing zeros until the 50th byte.

Reply:
So, I guess we should copy into length 51 char array, then do a `strlen` and then use [QString::fromLocal8Bit()](https://doc.qt.io/qt-6/qstring.html#fromLocal8Bit) with the length we have?

Reply:
That seemed to work.